File: Languages/Ruby/QuickSort.rb
#!/usr/bin/env ruby

# Source: ChatGPT

$LOGGING = false

def quicksort(arr)
  # Base case: an array of length 0 or 1 is already sorted
  return arr if arr.length <= 1

  # Choose a pivot element
  pivot = arr[0]

  # Divide the array into two sub-arrays of elements less than and greater than the pivot
  less = []
  greater = []
  arr[1..-1].each do |x|
    if x < pivot
      less << x
    else
      greater << x
    end
  end

  # Recursively sort the two sub-arrays and concatenate them with the pivot in between
  return quicksort(less) + [pivot] + quicksort(greater)
end

def main()
  arr = [10650,56341,77676,47158,34999,22009,19680,41870,31090,30668]
  ans = quicksort(arr)
  if $LOGGING
    puts "Sorted array: #{ans}"
  else
    puts ""
  end
end

main()

File: Languages/Rust/QuickSort.rs
// Source: ChatGPT

static mut LOGGING: bool = false;

fn main() {
    let arr = vec![5, 2, 6, 8, 1];
    quick_sort(&mut arr.clone(), 0, (arr.len() - 1) as i32);
    if unsafe { LOGGING } {
        println!("Sorted array: {:?}", arr); //Logging=True
    } 
}

fn quick_sort(arr: &mut [i32], lo: i32, hi: i32) {
    if lo < hi {
        let p = partition(arr, lo, hi);
        quick_sort(arr, lo, p - 1);
        quick_sort(arr, p + 1, hi);
    }
}

fn partition(arr: &mut [i32], lo: i32, hi: i32) -> i32 {
    let pivot = arr[lo as usize];
    let mut i = lo - 1;
    let mut j = hi + 1;
    loop {
        i += 1;
        while arr[i as usize] < pivot {
            i += 1;
        }
        j -= 1;
        while arr[j as usize] > pivot {
            j -= 1;
        }
        if i >= j {
            return j;
        }
        arr.swap(i as usize, j as usize);
    }
}
